File: DocumentAnalysisSystem/Models/DocumentComponent.cs
using System.Collections.Generic;
using DocumentAnalysisSystem.Visitors;

namespace DocumentAnalysisSystem.Models
{
    public abstract class DocumentComponent
    {
        protected DocumentComponent(string name)
        {
            Name = name;
        }

        public string Name { get; }

        public abstract IReadOnlyList<DocumentComponent> Children { get; }

        public abstract void Accept(IDocumentVisitor visitor);
    }
}

File: DocumentAnalysisSystemJava/src/main/java/com/documentanalysis/models/DocumentComponent.java
package com.documentanalysis.models;

import java.util.List;

public abstract class DocumentComponent {
    private final String name;

    protected DocumentComponent(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public abstract List<DocumentComponent> getChildren();

    public abstract void accept(com.documentanalysis.visitors.IDocumentVisitor visitor);
}
